src/libraries/Common/src/Interop/Browser/Interop.HttpHandlerService.cs

```diff
@@ -0,0 +1,535 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JSObject = Interop.JavaScript.JSObject;
+using JSException = Interop.JavaScript.JSException;
+using HostObject = Interop.JavaScript.HostObject;
+using Uint8Array = Interop.JavaScript.Uint8Array;
+using Function = Interop.JavaScript.Function;
+
+internal static partial class Interop
+{
+    internal static partial class Browser
+    {
+        internal interface IHttpHandlerService
+        {
+            Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
+            void Dispose();
+        }
+
+        public class BrowserHttpHandlerService : IHttpHandlerService, IDisposable
+        {
+
+            private static JSObject? fetch;
+            private static JSObject? window;
+
+            /// <summary>
+            /// Gets whether the current Browser supports streaming responses
+            /// </summary>
+            private static bool StreamingSupported { get; }
+
+            public BrowserHttpHandlerService()
+            { }
+
+            static BrowserHttpHandlerService()
+            {
+                using (var streamingSupported = new Function("return typeof Response !== 'undefined' && 'body' in Response.prototype && typeof ReadableStream === 'function'"))
+                    StreamingSupported = (bool)streamingSupported.Call();
+                handlerInit();
+            }
+
+            private static void handlerInit()
+            {
+                window = (JSObject)Interop.Runtime.GetGlobalObject("window");
+                fetch = (JSObject)Interop.Runtime.GetGlobalObject("fetch");
+            }
+
+            public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                // There is a race condition on Safari as a result of using TaskCompletionSource that
+                // causes a stack exceeded error being thrown.  More information can be found here:
+                // https://devblogs.microsoft.com/premier-developer/the-danger-of-taskcompletionsourcet-class/
+                var tcs = new TaskCompletionSource<HttpResponseMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
+                using (cancellationToken.Register(() => tcs.TrySetCanceled()))
+                {
+#pragma warning disable 4014
+                    doFetch(tcs, request, cancellationToken).ConfigureAwait(false);
+#pragma warning restore 4014
+
+                    return await tcs.Task.ConfigureAwait(true);
+                }
```

Comment:
I'm confused by this method.  doFetch both returns a Task and accepts a cancellationToken; why do we need to wrap it with another task like this?  Does doFetch not actually guarantee timely reaction to a cancellation request?  If that's the issue, this approach also means that we'll be signaling back to the caller via a canceled task that the operation completed as canceled, even if it's still running... that's generally not a good idea.

Reply:
Addressed in recent commit.